nrf_rpc/doc/usage.rst

```diff
@@ -0,0 +1,173 @@
+Group is created with the :c:macro:`NRF_RPC_GROUP_DEFINE`.
+Grouping allows logically divide the remote API, but also increase performance of nRF_PRC.
+
+
+Encoders
+========
+
+Encoders encodes commands and events into serialized packets.
+Creating an encoder is similar for all packet type.
+The first step is allocation of the buffer using e.g. :c:macro:`NRF_RPC_CMD_ALLOC`, :c:macro:`NRF_RPC_CBOR_EVT_ALLOC` or similar depending what kind of packet will be send.
+After that you can encode parameters directly into buffer or using `TinyCBOR <https://intel.github.io/tinycbor/current/>`_ library.
+In the last step packet is send using one of the sending function, e.g. :cpp:func:`nrf_rpc_cmd_send`, :cpp:func:`nrf_rpc_cbor_evt_send` or similar.
+
+As the result of sending command response is received, so it have to be parsed.
+There are two ways to prase the response.
+
+First is to provide response handler in parameters of :cpp:func:`nrf_rpc_cmd_send`.
+It will be called before :cpp:func:`nrf_rpc_cmd_send` returns.
+It may be called from a different thread.
+
+Second is to call :cpp:func:`nrf_rpc_cmd_rsp_send` or :cpp:func:`nrf_rpc_cbor_cmd_rsp_send` which have output parameters that will contain the response.
+After parsing it :cpp:func:`nrf_rpc_decoding_done` function must be called to indicate that parsing is done and the buffers holding the response can be released.
+
+Events have no response, so nothing more have to be done after sending it.
+
+Sample command encoder using TinyCBOR API:
+
+.. code-block:: c
+
+	/* Helper define holding maximum CBOR encoded packet length
+	 * for this sample.
+	 */
+	#define MAX_ENCODED_LEN 16
+
+	/* Defines a group that contains functions implemented in this
+	 * sample.
+	 */
+	NRF_RPC_GROUP_DEFINE(math_group, "sample_math")
+
+	/* Defines a helper structure to pass the results.
+	 */
+	struct remote_inc_result {
+		int err;
+		int output;
+	};
+
+	/* Function will remotely increment `input` by one and put the
+	 * result into `output`. Function returns 0 on success or
+	 * non-zero error code.
+	 */
+	int remote_inc(int input, int *output)
+	{
+		int err;
+		CborEncoder *encoder;
+		struct remote_inc_result result;
+		struct nrf_rpc_cbor_alloc_ctx ctx;
+
+		NRF_RPC_CBOR_CMD_ALLOC(ctx, &math_group, encoder,
+				       MAX_ENCODED_LEN,
+				       return NRF_RPC_NO_MEM);
+
+		cbor_encode_int(encoder, input);
+
+		err = nrf_rpc_cbor_cmd_send(&ctx, MATH_COMMAND_INC,
+					    remote_inc_rsp, &result);
+
+		if (err == 0) {
+			*output = result.output;
+			err = result.err;
+		}
+
+		return err;
+	}
+
+Above code uses `remote_inc_rsp` function to parse the response.
+Following code shows how this function may look like.
+
+.. code-block:: c
+
+	static int remote_inc_rsp(CborValue *parser, void *handler_data)
+	{
+		CborError cbor_err;
+		struct remote_inc_result *result =
+			(struct remote_inc_result *)handler_data;
+
+	 	if (!cbor_value_is_integer(parser)) {
+			goto cbor_error_exit;
```

Comment:
In an academic context using goto is not recommended unless it is justified.
Here it can be easily refactored out and this is what I am suggesting to do.

Reply:
Generally yes, but there are many disputes what `justified` means. In this context we already some clarification. In nrfxlib we are using [Zephyr Coding Style](https://docs.zephyrproject.org/latest/contribute/index.html#coding-style) for new modules if possible. It is basically Linux kernel coding style with some exceptions. See [Centralized exiting of functions](https://www.kernel.org/doc/html/latest/process/coding-style.html#centralized-exiting-of-functions).

Reply:
There are two exit points though...

Reply:
Yes and it is allowed. See example code from chapter that I linked before: [Centralized exiting of functions](https://www.kernel.org/doc/html/latest/process/coding-style.html#centralized-exiting-of-functions)

Reply:
No, you need to read the WHOLE paragraph of the chapter:
> The goto statement comes in handy when a function exits from multiple locations and **some common work such as cleanup has to be done. If there is no cleanup needed then just return directly**.

and no cleanup is done here.
This exactly show why `goto` should raise a red warning flag when found in code.
(I agree with kernel, that `goto` is useful when cleanup must be done, to exactly ensure ALL exit points does clean up.)

Reply:
Maybe my English language understanding is wrong, so correct me if I am wrong.

"**some common work such as cleanup has to be done**" means that some common work has to be done and an example of such common work is a cleanup, but **common work** is not limited to cleanup only. In our context our common work is setting correct common state into a structure.

Reply:
On the other hand this code is a part of the documentation. Some people may have very different opinion of `goto` (even pretty radical), so removing it from the documentation may be a good idea.

I will do that.

Reply:
The second sentence "**If there is no cleanup needed then just return directly**" may suggest that `cleanup` is a requirement, but actually intention was different. See the commit from 7 years ago when the sentence was added: https://github.com/torvalds/linux/commit/b57a0505e750b3d7b2d39e9823b276d8ca1a08fe

Reply:
I also looked at the linux kernel sources to see how they are understanding this chapter and I find similar constructs that I did, e.g.:
```c
	...
	if (!entry)
		goto exit_err;
	...
	...
	return entry;
exit_err:
	return ERR_PTR(err);
}
```

or:

```c
	...
		if (!btf_type_is_struct(elem_type))
			goto error;
	...
	...
		return ...;
error:
		bpf_log(log, "access beyond struct %s at off %u size %u\n",	tname, off, size);
		return -EACCES;
	}
```

Reply:
In these cases it easy to avoid to use goto, and it also makes the code easier to read:
```
cbor_err = cbor_value_get_int(value, &result->output);
if (cbor_err != CborNoError) {
	result->err = -EINVAL
	return;
}
```
When it is easy to avoid using structures that are considered to be bad practice, we should. We should not simply use gotos because it is allowed or because it is present in the linux kernel.

Reply:
> The second sentence "**If there is no cleanup needed then just return directly**" may suggest that `cleanup` is a requirement, but actually intention was different. See the commit from 7 years ago when the sentence was added: [torvalds/linux@b57a050](https://github.com/torvalds/linux/commit/b57a0505e750b3d7b2d39e9823b276d8ca1a08fe)

Well, if I read the commit message:
>  Documentation/CodingStyle: allow multiple return statements per function
> 
> A surprising number of newbies interpret this section to mean that only
one return statement is allowed per function.  Part of the problem is that
the "one return statement per function" rule is an actual style guideline
that people are used to from other projects.

it actually backs my claim that multiple returns are allowed, and `goto` should ONLY be used for cleanup.

The fact that you might find some places that don't honor this is just an indication that the guideline was not clear enough previously, or that reviewers haven't been strict on this rule when reviewing.
(You can also find places in Zephyr where coding guidelines are not followed)

Reply:
Yes, I agree that examples from linux kernel are not proofing anything, but I must disagree with your @tejlmand interpretation of the commit message. There is nothing regarding cleanup. The intention of this commit is to point out that "one return statement per function" is not a rule for this coding style.

Reply:
it's always hard to read other peoples mind.
And people may not always follow the rules they make themselves.

But in your code example I see no reason why a `goto` is justified.
Whereas, in those:

https://github.com/torvalds/linux/commit/7d8697afae5c61a9495fff54a1c5d053dea4505f
https://github.com/torvalds/linux/commit/3026050179a3a9a6f5c892c414b5e36ecf092081
some form of cleanup/dropped counter is incremented.

```
update_error_stats:
	u64_stats_update_begin(&txq->txq_stats.syncp);
	txq->txq_stats.tx_dropped++;
	u64_stats_update_end(&txq->txq_stats.syncp);
	return err;
}
```
and:
```
inc_dropped:
	dev->stats.tx_dropped++;
```

If I wanted to spend time of searching, there are thousand of places where code returns immediately without a `goto`.

Especially in documentation with a user example, there is no reason to use a `goto`, cause as this thread proofs, the use of `goto` is highly debatable, and thus should be avoided, unless there is a really good reason to use it.
You try to argue that the use of `goto` is allowed, but you haven't justified exactly why the `goto` is a better choice here, than simply do a return earlier.

Reply:
First, I agree that it is hard to read "what author had in mind", so I don't want to go further this path.

I already agreed that in this specific sample code in the documentation is better to use return. I changed it already few days ago. I was thinking that your comment in general discussion

> Remove the use of goto

was applying to entire PR. If I was wrong and you talking about this one specific `goto` then there is no point of continuing this discussion, because I already agreed and made the change.

If not and you are suggesting removing other `goto`s then I want to defend my code.

Reply:
> If not and you are suggesting removing other gotos then I want to defend my code.

I haven't noticed other `goto`s, but now you made me curios so will take a closer look.

Reply:
There is one `goto` in this PR: https://github.com/nrfconnect/sdk-nrfxlib/blob/3ecc0861e513faeb0b71597db07d1ca876ebfa5a/nrf_rpc/nrf_rpc.c#L484

I hope that you agree that it is justified usage.

The second that I was thinking is actually in the related PR: https://github.com/nrfconnect/sdk-nrf/blob/55f73a0826ea97fa0efcc75141e7ec970b083593/subsys/nrf_rpc/nrf_rpc_rpmsg.c#L115